### DISCUSSION TITLE
New Icon Engine

### DISCUSSION THREAD

**Discussion by vhyrro** [nvim-neorg/neorg#23] [Features]
# Prologue
Before we begin coding the meaty stuff like a reimagined `org-capture` etc. I thought it was important to flesh out the current user experience - icons.

# Features
`core.norg.concealer` just got a big overwrite in the `unstable` branch (it will come to `main` once extensively documented) - instead of using Neovim's `conceal` it now uses extmarks to display virtual text.

The implications of this are pretty cool:
- Coloured icons (each icon can be bound to an individual highlight group)
- Better headings (instead of concealing all asterisks from a heading into one icon we can hide away the first set of asterisks and "iconify" the last one) - this makes looking at non-indented headings much more pleasant
- More flexible - you can practically overlay any set of text over any other set of text

# Screenshots
![example-todo](https://user-images.githubusercontent.com/76052559/125930409-1c654bc8-316c-41ef-b0cd-d71462dd544f.png)
![markers-and-quotes](https://user-images.githubusercontent.com/76052559/125930599-1e0f8326-bc94-40fe-992e-f4694c113487.png)
![headings](https://user-images.githubusercontent.com/76052559/125931024-5c43eaf2-e82b-4ddb-bfba-47de975f35d1.png)

I'm open to discussion on the default icon choice. I don't know about the flower (I just don't have a better looking alternative right now).

**Comment by akinsho** [comment]
@vhyrro glad to see you got it working 👍🏿 . I think the default icon set that org bullets use are nice enough as long as users can change them as they go. I don't think it's configurable in org mode yet people have just accepted that so I don't think it needs too much fussing. In my experience the nvim community like to tweak a whole bunch more so I think exactly what the defaults are won't matter *too* much as long as people can change things.

**Reply by vhyrro** [reply]
As with pretty much everything in Neorg, yes, the icons are configurable, I just wanna make it nice OOTB so people don't need to go out of their way to change stuff :)

It's also possible to add your own custom icons with your own regex patterns and parameters. I just haven't written wiki entries for that yet, which is what I'm about to do. Thanks once again for letting me know about the extmark virtual text stuff, it'll come in really handy!

**Comment by mrossinek** [comment]
> I don't know about the flower (I just don't have a better looking alternative right now).

How about something like the following:
```lua
["core.norg.concealer"] = {
            config = {
                icons = {
                    heading = {
                        level_1 = {
                            icon = "◉",
                            },
                        level_2 = {
                            icon = "◎",
                            },
                        level_3 = {
                            icon = "○",
                            },
                        level_4 = {
                            icon = "∙",
                            },
                        },
                    },
                },
            },
```
This will look similar to this (don't mind the highlighting right now...)
![screenshot_1627679027](https://user-images.githubusercontent.com/21973473/127711344-f26cea5b-79db-4941-b9c2-c0e52ea68ccb.png)

The changes above are:
- use the current level 2 icon for the third level
- replace the flower with https://unicode-table.com/en/25CE/
- replace the level 4 icon (called "bullet") with the mathematical bullet operator (which appears to be more centered) (https://unicode-table.com/en/2219/)

In general I found these icons a good inspiration: https://unicode-table.com/en/blocks/geometric-shapes/

**Reply by vhyrro** [reply]
Sounds great, I'll check em out when i have the time. Thanks!

**Reply by vhyrro** [reply]
Man, they look waay nicer on your machine lol

On my end the second heading is massive haha. The second and third element seem very close in size and so it's a bit hard to read/look at. Fonts are hard :/

**Reply by mrossinek** [reply]
Hm interesting :thinking: FWIW here are the fonts I use:
- main: [Meslo Nerd Font](https://github.com/ryanoasis/nerd-fonts/tree/master/patched-fonts/Meslo/S)
- [font awesome](https://fontawesome.com/) icons (I think these are included in the patched Nerd font but for some reason I have them as a separate file as well :shrug:)
- additional icons: [nonicons](https://github.com/yamatsum/nonicons)
- additional weather icons: https://erikflowers.github.io/weather-icons/
